ocaml/xapi/smint.ml

```diff
@@ -131,30 +132,57 @@ let parse_string_int64_features features =
     | feature :: _ ->
         error "SM.feature: unknown feature %s" feature ;
         None
-  in
-  features
-  |> List.filter_map scan
-  |> List.sort_uniq (fun (x, _) (y, _) -> compare x y)
 
-(** [compat_features features1 features2] finds the compatible features in the input
-features lists. We assume features backwards compatible, i.e. if there are FOO/1 and
+  (** [compat_features features1 features2] finds the compatible features in the input
+  features lists. We assume features backwards compatible, i.e. if there are FOO/1 and
   FOO/2 are present, then we assume they can both do FOO/1*)
-let compat_features features1 features2 =
-  let features2 = List.to_seq features2 |> Hashtbl.of_seq in
-  List.filter_map
-    (fun (f1, v1) ->
-      match Hashtbl.find_opt features2 f1 with
-      | Some v2 ->
-          Some (f1, Int64.min v1 v2)
-      | None ->
-          None
-    )
-    features1
-
-let parse_capability_int64_features strings =
-  List.map
-    (function c, v -> (List.assoc c string_to_capability_table, v))
-    (parse_string_int64_features strings)
+  let compat_features features1 features2 =
+    let features2 = List.to_seq features2 |> Hashtbl.of_seq in
+    List.filter_map
+      (fun (f1, v1) ->
+        match Hashtbl.find_opt features2 f1 with
+        | Some v2 ->
+            Some (f1, Int64.min v1 v2)
+        | None ->
+            None
+      )
+      features1
+
+  let of_string_int64_opt (c, v) =
+    List.assoc_opt c string_to_capability_table |> Option.map (fun c -> (c, v))
+
+  (** [has_capability c fl] will test weather the required capability [c] is present 
+  in the feature list [fl]. Callers should use this function to test if a feature
+  is available rather than directly using membership functions on a feature list
+  as this function might have special logic for some features. *)
+  let has_capability (c : capability) (fl : t list) =
+    List.exists
+      (fun (c', _v) ->
+        match (c, c') with
+        | Vdi_mirror_in, Vdi_mirror ->
+            true
+        | c, c' when c = c' ->
+            true
+        | _ ->
+            false
+      )
+      fl
+
+  (** [parse_string_int64 features] takes a [features] list in its plain string
```

Comment:
I was expecting some code that splits a string into two parts but did not see it.

Reply:
Yes, see `string_int64_of_string_opt`